### PR TITLE
feat: implement new controllers for hostname, resolvers and time servers

### PR DESCRIPTION
**Pull request by smira**
This is next part of networkd rewrite.

This implements three new resource types coupled with controllers which
process the default configuration, merges and applying changes.

TimeSync was set up to watch the time servers resource. This is a no-op
for now, but once DHCP is implemented, this would enable time server
configuration coming from DHCP.

Signed-off-by: Andrey Smirnov <smirnov.andrey@gmail.com>

### CODE REVIEW

File: internal/app/machined/pkg/controllers/network/hostname_config.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+func (ctrl *HostnameConfigController) parseMachineConfiguration(logger *zap.Logger, cfgProvider talosconfig.Provider) (spec network.HostnameSpecSpec) {
+	hostname := cfgProvider.Machine().Network().Hostname()
+
+	if hostname == "" {
+		return
+	}
+
+	if err := spec.ParseFQDN(hostname); err != nil {
+		logger.Info("ignoring error", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (smira) [reply internal/app/machined/pkg/controllers/network/hostname_config.go:221]: good point, yeah, probably warning is better :+1: 

I don't want controller to fail completely on a single error, but rather try to do _something_

File: pkg/resources/network/hostname_spec.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	spec.Hostname = parts[0]
+
+	if len(parts) > 1 {
+		spec.Domainname = parts[1]
+	}
+
+	return spec.Validate()
+}
+
+// NewHostnameSpec initializes a SecretsStatus resource.
```
Comment (Unix4ever) [inline pkg/resources/network/hostname_spec.go:71]:
```suggestion
// NewHostnameSpec initializes a Hostname resource.
```

File: pkg/resources/network/hostname_status.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+// FQDN returns the fully-qualified domain name.
+func (spec *HostnameStatusSpec) FQDN() string {
+	if spec.Domainname == "" {
+		return spec.Hostname
+	}
+
+	return spec.Hostname + "." + spec.Domainname
+}
+
+// NewHostnameStatus initializes a SecretsStatus resource.
```
Comment (Unix4ever) [inline pkg/resources/network/hostname_status.go:38]:
```suggestion
// NewHostnameStatus initializes a HostnameStatus resource.
```

And it looks like all other resources have the same copy-paste error ;)

**Review by Unix4ever — APPROVED** [review]
:+1: with small nits

File: pkg/resources/network/hostname_status.go (hunk 3 above)
Comment (smira) [reply pkg/resources/network/hostname_status.go]: thanks, fixed all of them!

**Review by Unix4ever — APPROVED** [review]

**Pull request merged**